pymongo/mongo_client.py

```diff
@@ -1217,30 +1199,51 @@ def _send_message_with_response(self, operation, exhaust=False,
             and server.description.server_type != SERVER_TYPE.Mongos) or (
                 operation.read_preference != ReadPreference.PRIMARY)
 
-        return self._reset_on_error(
-            server,
-            server.send_message_with_response,
-            operation,
-            set_slave_ok,
-            self.__all_credentials,
-            self._event_listeners,
-            exhaust)
-
-    def _reset_on_error(self, server, func, *args, **kwargs):
-        """Execute an operation. Reset the server on network error.
+        with self._reset_on_error(server.description.address,
+                                  operation.session):
+            return server.send_message_with_response(
+                operation,
+                set_slave_ok,
+                self.__all_credentials,
+                self._event_listeners,
+                exhaust)
 
-        Returns fn()'s return value on success. On error, clears the server's
-        pool and marks the server Unknown.
+    @contextlib.contextmanager
+    def _reset_on_error(self, server_address, session):
+        """Reset the server according to SDAM errors.
```

Comment:
Can you add a comment here mentioning which "SDAM errors" this docstring is referring to? 
In the [specification](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#error-handling), errors (and the drivers' handling of them) are categorized and explained in terms of what the driver was trying to do when the error was encountered. It is not immediately clear what part of the spec is being implemented by this helper.

Reply:
Updated to mention "not master" or "node is recovering" errors from the SDAM spec.